compatibility/bazel_tools/create-daml-app/Main.hs

```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
         [ test getTools
         ]
   where
-    test getTools = testCaseSteps "Getting Started Guide" $ \step -> withTempDir $ \tmpDir -> do
+    test getTools = testCaseSteps "Getting Started Guide" $ \step -> withTempDir $ \tmpDir' -> do
+        -- npm gets confused when the temporary directory is not under 'private' on mac.
+        let tmpDir
+              | isMac = "/private" <> tmpDir'
+              | otherwise = tmpDir'
         Tools{..} <- getTools
         setEnv "CI" "yes" True
         step "Create app from template"
         withCurrentDirectory tmpDir $ do
           callProcess damlBinary ["new", "create-daml-app", "create-daml-app"]
         let cdaDir = tmpDir </> "create-daml-app"
+        let uiDir = cdaDir </> "ui"
         step "Patch the application code with messaging feature"
         withCurrentDirectory cdaDir $ do
           callProcessSilent patchPath ["-p2", "-i", messagingPatch]
           forM_ ["MessageEdit", "MessageList"] $ \messageComponent ->
             assertFileExists ("ui" </> "src" </> "components" </> messageComponent <.> "tsx")
         step "Extract codegen output"
-        runConduitRes
-            $ Conduit.sourceFile codegenPath
-            .| Zlib.ungzip
-            .| Tar.untar (restoreFile (\a b -> fail (T.unpack $ a <> b)) (cdaDir </> "daml.js"))
-        withCurrentDirectory (cdaDir </> "ui") $ do
+        extractTarGz codegenPath $ uiDir </> "daml.js"
+        -- we patch all the 'package.json' files to point to the local versions of the TypeScript
+        -- libraries.
+        genFiles <- listFilesRecursive $ uiDir </> "daml.js"
+        forM_ [file | file <- genFiles, takeFileName file == "package.json"] (patchTsDependencies uiDir)
+        withCurrentDirectory uiDir $ do
           step "Set up libraries and workspaces"
-          setupYarnEnv tmpDir (Workspaces ["create-daml-app/ui"])
-            [(DamlLedger, damlLedgerPath), (DamlReact, damlReactPath), (DamlTypes, damlTypesPath)]
+          setupNpmEnv uiDir [(DamlTypes, damlTypesPath)
+                            , (DamlLedger, damlLedgerPath)
+                            , (DamlReact, damlReactPath)
+                            ]
           step "Install Jest, Puppeteer and other dependencies"
           addTestDependencies "package.json" testDepsPath
-          retry 3 (callProcessSilent yarnPath ["install"])
+          patchTsDependencies uiDir "package.json"
+          -- use '--scripts-prepend-node-path' to make sure we are using the correct 'node' binary
+          retry 3 (callProcessSilent npmPath ["install", "--scripts-prepend-node-path"])
           step "Run Puppeteer end-to-end tests"
-          copyFile testTsPath (cdaDir </> "ui" </> "src" </> "index.test.ts")
-          callProcess yarnPath ["run", "test", "--ci", "--all"]
+          copyFile testTsPath (uiDir </> "src" </> "index.test.ts")
+          -- we need 'npm-cli.js' in the path for the following test
```

Comment:
Why do we need this in this test but I don’t need seem to have it in PATH on my machine where I’ve installed NPM?

Reply:
you won't have 'npm-cli.js' in your PATH because it's a wrapper from bazel for `npm`.